EXR support for snmp_community

tphoney opened the pull request:
Hi @mikewiebe @chburket, as discussed from today. Issues:
- a group is not required to be associated with a SNMP community.
- I am very unclear about how associate an snmp group with a snmp community. 
- Also how to create an ACL that can be associated with an snmp community.
- the test methodology seems dated, i tried my best to match what was there.

Can you provide feedback on this ? or any other feedback ?

Commit:
EXR support for snmp_community

tphoney commented:
please note, with out acl and group, the provider is kinda simple.

mikewiebe commented:
@tphoney I have reached out to the XR snmp development team to get some clarity on these issues.  I will summarize here once I have a chance to discuss this with them but it might be a bit delayed as they are in a different time zone (India).

mikewiebe commented:
@tphoney Quick update here.  The developer responsible for this feature is currently on leave so I have asked for another domain expert to help weigh in.

tphoney commented:
@mikewiebe thanks for the update

mikewiebe commented:
@tphoney Finally was able to get some time with the XR snmp development team this morning and have more clarity on how this works on XR.  I need to test a few more things but will add a short writeup later today.

tphoney commented:
➜  cisco-network-node-utils git:(snmp_community) ✗ bundle exec ruby tests/test_snmpcommunity.rb -v -- j2nvya0j6ll9jel.delivery.puppetlabs.net:57777 root lab
Run options: -v -- --seed 50464

# Running:

Node under test:
- name  - j2nvya0j6ll9jel
- type  - R-IOSXRV9000-CH
- image - 6.0.0.20I

TestSnmpCommunity#test_snmpcommunity_acl_set_zero_length = 33.78 s = .
TestSnmpCommunity#test_snmpcommunity_group_set_zero_length = 32.47 s = .
TestSnmpCommunity#test_snmpcommunity_acl_set_default = 33.74 s = .
TestSnmpCommunity#test_snmpcommunity_acl_set_valid = 27.81 s = .
TestSnmpCommunity#test_snmpcommunity_create_with_name_alphanumeric_char = 25.25 s = .
TestSnmpCommunity#test_snmpcommunity_create_group_zero_length = 10.22 s = .
TestSnmpCommunity#test_snmpcommunity_group_set_default = 58.66 s = .
TestSnmpCommunity#test_snmpcommunity_acl_get_no_acl = 46.31 s = .
TestSnmpCommunity#test_snmpcommunity_group_set_too_long = 8.79 s = .
TestSnmpCommunity#test_snmpcommunity_create_name_too_long = 7.95 s = .
TestSnmpCommunity#test_snmpcommunity_create_group_nil = 6.49 s = .
TestSnmpCommunity#test_snmpcommunity_group_set_valid = 43.33 s = .
TestSnmpCommunity#test_snmpcommunity_collection_not_empty = 36.33 s = .
TestSnmpCommunity#test_snmpcommunity_acl_get = 56.75 s = .
TestSnmpCommunity#test_snmpcommunity_acl_set_nil = 32.95 s = .
TestSnmpCommunity#test_snmpcommunity_collection_empty = 17.88 s = .
TestSnmpCommunity#test_snmpcommunity_group_set_invalid = 6.85 s = .
TestSnmpCommunity#test_snmpcommunity_create_name_nil = 5.70 s = .
TestSnmpCommunity#test_snmpcommunity_collection_valid = 35.54 s = .
TestSnmpCommunity#test_snmpcommunity_create_valid = 59.03 s = .
TestSnmpCommunity#test_snmpcommunity_create_group_invalid = 5.49 s = .
TestSnmpCommunity#test_snmpcommunity_create_name_zero_length = 6.14 s = .
TestSnmpCommunity#test_snmpcommunity_destroy_valid = 25.41 s = .
TestSnmpCommunity#test_snmpcommunity_create_group_too_long = 12.31 s = .
TestSnmpCommunity#test_snmpcommunity_get_group = 60.42 s = .

Finished in 703.210927s, 0.0356 runs/s, 0.1166 assertions/s.

25 runs, 82 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /Users/tp/workspace/git/EXR/cisco_devenv/cisco-network-node-utils/coverage. 2337 / 4563 LOC (51.22%) covered.

mikewiebe commented on lib/cisco_node_utils/snmpcommunity.rb:
```diff
@@ -22,10 +22,23 @@ class SnmpCommunity < NodeUtil
+                   state: '',
+                   name:  @name)
+        # create the mapping for group
+        config_set('snmp_community', 'group_simple', state: '', group: group)
```
I don't see `state` in the yaml file for the `group_community_mapping` property.

mikewiebe commented on lib/cisco_node_utils/snmpcommunity.rb:
```diff
@@ -41,7 +54,10 @@ def self.communities
+      config_set('snmp_community', 'community',
+                 state: 'no',
+                 name:  @name,
+                 group: 'null')
```
`null` works for both nxos and xr?

mikewiebe commented on lib/cisco_node_utils/snmpcommunity.rb:
```diff
@@ -50,12 +66,22 @@ def destroy
     def group
-      config_get('snmp_community', 'group', @name)
+      if platform == :nexus
+        config_get('snmp_community', 'group', name: @name)
```
I am fairly certain this will break nxos.  You are using printf style wildcards in the yaml file but key-value wildcards when calling `config_get`.

https://github.com/cisco/cisco-network-node-utils/pull/315/files#diff-3e8591eaf966384e60ee4ec16e769fd8R38

Please update the yaml file to support key-value wildcards across all properties instead of a mix like you have now.

chburket commented on tests/test_snmpcommunity.rb:
```diff
@@ -34,45 +34,87 @@ class TestSnmpCommunity < CiscoTestCase
+        s = md.post_match
+      end
+    else
+      config('no snmp-server community com1',
```
Is there a way to do something more dynamic like we do for nexus?  Won't this test fail for XR if an snmp community is defined that is not one of the explicitly listed ones?

mikewiebe replied:
This cleanup section won't be very reliable if other communities are setup on the test device.  I would prefer you use logic similar to nexus above where you use the 'show running' command go get all existing communities and remove them or use the class method to gather all communities and destroy them in the setup.

mikewiebe replied:
Why not just call `cleanup_snmp_communities(snmpcommunities)` here for both nxos and xr?

chburket commented on tests/test_snmpcommunity.rb:
```diff
@@ -120,6 +166,7 @@ def test_snmpcommunity_create_group_too_long
   def test_snmpcommunity_create_group_invalid
     name = 'ciscotest'
     group = 'network-operator-invalid'
+    return if platform != :nexus
```
We should either make this test work for XR or explicitly skip it.

chburket commented on tests/test_snmpcommunity.rb:
```diff
@@ -162,6 +227,7 @@ def test_snmpcommunity_group_set_zero_length
   end
 
   def test_snmpcommunity_group_set_too_long
+    return if platform != :nexus
```
Skip on XR or make this test work?

chburket commented on tests/test_snmpcommunity.rb:
```diff
@@ -188,8 +255,18 @@ def test_snmpcommunity_group_set_valid
     snmpcommunity.group = group
-    assert_show_match(
-      pattern: /snmp-server community\s#{name}\sgroup\s#{group}/)
+    if platform != :ios_xr
```
nit... this pattern is repeated ~5 times.  Consider creating a helper method for verifying the existence of a community by name via "show" output.

tphoney commented:
@mikewiebe @chburket Hi ! long time no see. I am back from vacation so i have had a chance to look at this again. I have taken each of your comments in turn
- I don't see state in the yaml file for the group_community_mapping property. removed state from yaml, un-needed.
- null works for both nxos and xr? i just changed the formatting, it works. 
- I am fairly certain this will break nxos. You are using printf style wildcards in the yaml file but key-value wildcards when calling config_get. fixed throughout
- cleaned up test_snmpcommunity_collection_empty return if platform != :nexus becomes return if platform == :ios_xr
- nit... this pattern is repeated ~5 times. Consider creating a helper method for verifying the existence of a community by name via "show" output. not sure what you mean by this

25 runs, 82 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest to /Users/tp/workspace/git/EXR/cisco_devenv/cisco-network-node-utils/coverage. 2424 / 4975 LOC (48.72%) covered.

tphoney commented:
i will use skip instead of return based on a conversation with @chburket

tphoney commented:
@chburket i made the suggested change of adding a skip, instead of a return. Ready for re-review.

chburket commented:
:+1:

tphoney commented:
Cool, can we get this merged ? @chburket

mikewiebe commented:
👍 I ran the tests on n3k(I2, I3), n5k, n6k, n7k, n8k and n9k(I2, I4).  All pass.
